src/core/ddsi/src/q_transmit.c

```diff
@@ -1024,6 +1068,117 @@ static int maybe_grow_whc (struct writer *wr)
   return 0;
 }
 
+int write_sample_p2p_wrlock_held(struct writer *wr, seqno_t seq, struct nn_plist *plist, struct ddsi_serdata *serdata, struct ddsi_tkmap_instance *tk, struct proxy_reader *prd)
+{
+  struct q_globals * const gv = wr->e.gv;
+  int r;
+  nn_mtime_t tnow;
+  int rexmit = 1;
+  struct wr_prd_match *wprd = NULL;
+  seqno_t gseq;
+  struct nn_xmsg *gap = NULL;
+  seqno_t gapstart = -1, gapend = -1, segstart;
+  unsigned gapbits[256 / 32];
+  unsigned gapnumbits = 0;
+
+  tnow = now_mt ();
+  serdata->twrite = tnow;
+  serdata->timestamp = now();
+
+  if (prd->filter)
+  {
+    if ((wprd = ddsrt_avl_lookup (&wr_readers_treedef, &wr->readers, &prd->e.guid)) != NULL)
+    {
+      rexmit = prd->filter(wr, prd, serdata);
+      /* determine if gap has to added */
+      if (rexmit)
+      {
+        GVLOG (DDS_LC_DISCOVERY, "send filtered "PGUIDFMT" last_seq=%"PRIu64" seq=%"PRIu64"\n", PGUID (wr->e.guid), wprd->seq, seq);
+
+        segstart = wprd->seq + 1;
+        memset (gapbits, 0, sizeof (gapbits));
+        for (gseq = segstart; gseq < seq; gseq++)
+        {
+          struct whc_borrowed_sample sample;
+          if (whc_borrow_sample (wr->whc, seq, &sample))
+          {
+            if (prd->filter(wr, prd, sample.serdata) == 0)
+            {
+              if (gapstart == -1)
```

Comment:
See also the comment in ``handle_AckNack``: three copies of the same logic really is too much. It would make sense to combine ``gapstart``, ``gapend``, ``gapnumbits`` and ``gapbits`` in a single struct and create a function to update that struct with following this logic. That can then be used in all call sites.